File: DitLog/DitLog/Models/Spot.swift
import Foundation

/// Unified spot from POTA or SOTA. In-memory only, not persisted.
struct Spot: Identifiable, Equatable {
    enum Source: String {
        case pota, sota
    }

    let id: String
    let activatorCallsign: String
    let frequency: Double        // MHz
    let mode: String

    // Dual reference fields — a spot can have both POTA and SOTA refs (after consolidation)
    var potaReference: String?
    var potaReferenceName: String?
    var sotaReference: String?
    var sotaReferenceName: String?

    let spotterCallsign: String?
    let comments: String?
    let timestamp: Date

    /// Which sources contributed to this spot
    var sources: Set<Source> {
        var s = Set<Source>()
        if potaReference != nil { s.insert(.pota) }
        if sotaReference != nil { s.insert(.sota) }
        return s
    }

    /// Primary source (backwards compat)
    var source: Source {
        if potaReference != nil { return .pota }
        return .sota
    }

    /// Primary reference (backwards compat)
    var reference: String {
        potaReference ?? sotaReference ?? ""
    }

    /// Primary reference name (backwards compat)
    var referenceName: String? {
        potaReferenceName ?? sotaReferenceName
    }

    /// Whether this spot is older than the expiry threshold
    func isExpired(after minutes: Double = 10) -> Bool {
        Date().timeIntervalSince(timestamp) > minutes * 60
    }

    /// Whether comments contain "QRT"
    var isQRT: Bool {
        guard let comments = comments else { return false }
        return comments.uppercased().contains("QRT")
    }

    /// The band derived from frequency
    var band: String {
        BandPlan.band(for: frequency) ?? "?"
    }

    /// Age of the spot in minutes (minimum 1 — a visible spot is never truly 0m old)
    var ageMinutes: Int {
        max(1, Int(Date().timeIntervalSince(timestamp) / 60))
    }
}

File: android/app/src/main/kotlin/com/sotalog/android/domain/models/Spot.kt
package com.sotalog.android.domain.models

import com.sotalog.android.domain.services.BandPlan
import java.util.Date

data class Spot(
    val id: String,
    val activatorCallsign: String,
    val frequency: Double,
    val mode: String,
    val potaReference: String? = null,
    val potaReferenceName: String? = null,
    val sotaReference: String? = null,
    val sotaReferenceName: String? = null,
    val spotterCallsign: String? = null,
    val comments: String? = null,
    val timestamp: Date,
) {
    enum class Source(val value: String) {
        POTA("pota"),
        SOTA("sota"),
    }

    val sources: Set<Source>
        get() = buildSet {
            if (potaReference != null) add(Source.POTA)
            if (sotaReference != null) add(Source.SOTA)
        }

    val source: Source
        get() = if (potaReference != null) Source.POTA else Source.SOTA

    val reference: String
        get() = potaReference ?: sotaReference ?: ""

    val referenceName: String?
        get() = potaReferenceName ?: sotaReferenceName

    fun isExpired(afterMinutes: Double = 60.0): Boolean {
        return (Date().time - timestamp.time) > (afterMinutes * 60 * 1000).toLong()
    }

    val isQRT: Boolean
        get() = comments?.uppercase()?.contains("QRT") == true

    val band: String
        get() = BandPlan.band(frequency) ?: "?"

    val ageMinutes: Int
        get() = maxOf(1, ((Date().time - timestamp.time) / 60_000).toInt())
}
